core/shared/src/main/scala/zio/ZLayer.scala

```diff
@@ -49,47 +61,79 @@ final case class ZLayer[-RIn, +E, +ROut <: Has[_]](value: ZManaged[RIn, E, ROut]
   def ++[E1 >: E, RIn2, ROut1 >: ROut <: Has[_], ROut2 <: Has[_]](
     that: ZLayer[RIn2, E1, ROut2]
   )(implicit tagged: Tagged[ROut2]): ZLayer[RIn with RIn2, E1, ROut1 with ROut2] =
-    ZLayer(
-      ZManaged.accessManaged[RIn with RIn2] { env =>
-        (self.value.provide(env) zipWith that.value.provide(env))((l, r) => (l: ROut1).union[ROut2](r))
+    new ZLayer(
+      Managed.finalizerRef(_ => UIO.unit).map { finalizerRef => memoMap =>
+        for {
+          l <- memoMap.getOrElseMemoize(self, finalizerRef)
+          r <- memoMap.getOrElseMemoize(that, finalizerRef)
+        } yield l.union[ROut2](r)
       }
     )
 
   def +!+[E1 >: E, RIn2, ROut2 <: Has[_]](
     that: ZLayer[RIn2, E1, ROut2]
   ): ZLayer[RIn with RIn2, E1, ROut with ROut2] =
-    ZLayer(
-      ZManaged.accessManaged[RIn with RIn2] { env =>
-        (self.value.provide(env) zipWith that.value.provide(env))((l, r) => l.unionAll[ROut2](r))
+    new ZLayer(
+      Managed.finalizerRef(_ => UIO.unit).map { finalizerRef => memoMap =>
+        for {
+          l <- memoMap.getOrElseMemoize(self, finalizerRef)
+          r <- memoMap.getOrElseMemoize(that, finalizerRef)
+        } yield l.unionAll[ROut2](r)
       }
     )
 
   /**
-   * Builds a layer that has no dependencies into a managed value.
+   * Builds a layer into a managed value.
+   */
+  def build: ZManaged[RIn, E, ROut] =
+    for {
+      memoMap <- ZLayer.MemoMap.make.toManaged_
+      run     <- self.scope
+      value   <- run(memoMap)
+    } yield value
+
+  /**
+   * Creates a fresh version of this layer that will not be shared.
    */
-  def build[RIn2 <: RIn](implicit ev: Any =:= RIn2): Managed[E, ROut] = value.provide(ev(()))
+  def fresh: ZLayer[RIn, E, ROut] =
+    new ZLayer(self.scope)
+
+  /**
+   * Returns an effect that, if evaluated, will return the lazily computed
+   * result of this layer.
+   */
+  def memoize: ZManaged[RIn, Nothing, ZLayer[Any, E, ROut]] =
```

Comment:
Is this useful anymore? I think we can delete it unless you have a use case in mind.

Reply:
Yes I think that is right.